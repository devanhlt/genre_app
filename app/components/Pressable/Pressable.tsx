import React, { ComponentType, ReactNode } from "react"
import {
  PressableProps,
  PressableStateCallbackType,
  Pressable as RNPressable,
  StyleProp,
  TextStyle,
  View,
  ViewStyle,
} from "react-native"

import { appColors, spacing, typography } from "app/theme"
import { IconTypes, IconWeight } from "../Icon/PhosphorIcon"
import { TypographyProps } from "../Typography"

const $baseViewStyle: ViewStyle = {
  paddingVertical: spacing.size12,
  paddingHorizontal: spacing.size16,
}

const $viewPresets = {
  primary: {
    medium: [
      $baseViewStyle,
      { backgroundColor: appColors.components.button.primary },
    ] as StyleProp<ViewStyle>,
    small: [
      $baseViewStyle,
      {
        backgroundColor: appColors.components.button.primary,
        paddingVertical: spacing.size08,
        paddingHorizontal: spacing.size16,
      },
    ] as StyleProp<ViewStyle>,
  },
  secondary: {
    medium: [
      $baseViewStyle,
      {
        backgroundColor: appColors.components.button.secondary,
        borderWidth: 1,
        borderColor: appColors.common.borderDefault,
      },
    ] as StyleProp<ViewStyle>,
    small: [
      $baseViewStyle,
      {
        backgroundColor: appColors.components.button.secondary,
        borderWidth: 1,
        borderColor: appColors.common.borderDefault,
        paddingVertical: spacing.size08,
        paddingHorizontal: spacing.size12,
      },
    ] as StyleProp<ViewStyle>,
  },
  tertiary: {
    medium: [
      $baseViewStyle,
      { backgroundColor: appColors.palette.transparent },
    ] as StyleProp<ViewStyle>,
    small: [
      $baseViewStyle,
      {
        backgroundColor: appColors.palette.transparent,
        paddingVertical: spacing.size08,
        paddingHorizontal: spacing.size12,
      },
    ] as StyleProp<ViewStyle>,
  },
}

type Presets = keyof typeof $viewPresets

const $viewOnlyIconPresets = {
  primary: {
    medium: [
      $baseViewStyle,
      {
        backgroundColor: appColors.components.button.primary,
        paddingVertical: spacing.size12,
        paddingHorizontal: spacing.size12,
      },
    ] as StyleProp<ViewStyle>,
    small: [
      $baseViewStyle,
      {
        backgroundColor: appColors.components.button.primary,
        paddingVertical: spacing.size08,
        paddingHorizontal: spacing.size08,
      },
    ] as StyleProp<ViewStyle>,
  },
  secondary: {
    medium: [
      $baseViewStyle,
      {
        backgroundColor: appColors.components.button.secondary,
        borderWidth: 1,
        borderColor: appColors.common.borderDefault,
        paddingVertical: spacing.size12,
        paddingHorizontal: spacing.size12,
      },
    ] as StyleProp<ViewStyle>,
    small: [
      $baseViewStyle,
      {
        backgroundColor: appColors.components.button.secondary,
        borderWidth: 1,
        borderColor: appColors.common.borderDefault,
        paddingVertical: spacing.size08,
        paddingHorizontal: spacing.size08,
      },
    ] as StyleProp<ViewStyle>,
  },
  tertiary: {
    medium: [
      $baseViewStyle,
      {
        backgroundColor: appColors.palette.transparent,
        paddingVertical: spacing.size12,
        paddingHorizontal: spacing.size12,
      },
    ] as StyleProp<ViewStyle>,
    small: [
      $baseViewStyle,
      {
        backgroundColor: appColors.palette.transparent,
        paddingVertical: spacing.size08,
        paddingHorizontal: spacing.size08,
      },
    ] as StyleProp<ViewStyle>,
  },
}

const $baseTextStyle: TextStyle = {
  textAlign: "center",
  flexShrink: 1,
  flexGrow: 0,
  zIndex: 2,
  marginStart: spacing.size04,
}

const $textPresets: Record<Presets, StyleProp<TextStyle>> = {
  primary: [$baseTextStyle, typography.label01, { color: appColors.common.characterWhite }],
  secondary: [$baseTextStyle, typography.label01, { color: appColors.common.characterRedDefault }],
  tertiary: [$baseTextStyle, typography.label01, { color: appColors.common.characterRedDefault }],
}

const $textDisabledPresets: Record<Presets, StyleProp<TextStyle>> = {
  primary: [$baseTextStyle, { color: appColors.common.characterDisabled }],
  secondary: [$baseTextStyle, { color: appColors.common.characterDisabled }],
  tertiary: [$baseTextStyle, { color: appColors.common.characterDisabled }],
}

const $pressedViewPresets: Record<Presets, StyleProp<ViewStyle>> = {
  primary: { backgroundColor: appColors.components.button.primaryPress },
  secondary: {
    backgroundColor: appColors.components.button.secondary,
    borderColor: appColors.common.borderPress,
  },
  tertiary: { backgroundColor: appColors.palette.transparent },
}

const $pressedViewDisabledPresets: Record<Presets, StyleProp<ViewStyle>> = {
  primary: { backgroundColor: appColors.components.button.disabled },
  secondary: {
    backgroundColor: appColors.components.button.secondary,
    borderColor: appColors.common.borderDisabled,
  },
  tertiary: { backgroundColor: appColors.palette.transparent },
}

const $pressedTextPresets: Record<Presets, StyleProp<TextStyle>> = {
  primary: { opacity: 1 },
  secondary: { color: appColors.common.characterBluePress },
  tertiary: { color: appColors.common.characterBluePress },
}

export interface ButtonAccessoryProps {
  style: StyleProp<any>
  pressableState: PressableStateCallbackType
  disabled?: boolean
  color?: string
}

export interface ButtonProps extends PressableProps {
  /**
   * Text which is looked up via i18n.
   */
  tx?: TypographyProps["tx"]
  /**
   * The text to display if not using `tx` or nested components.
   */
  text?: TypographyProps["text"]
  /**
   * Optional options to pass to i18n. Useful for interpolation
   * as well as explicitly setting locale or translation fallbacks.
   */
  txOptions?: TypographyProps["txOptions"]
  /**
   * An optional style override useful for padding & margin.
   */
  style?: StyleProp<ViewStyle>
  /**
   * An optional style override for the "pressed" state.
   */
  pressedStyle?: StyleProp<ViewStyle>
  /**
   * An optional style override for the button text.
   */
  viewStyle?: StyleProp<TextStyle>
  /**
   * An optional style override for the button text when in the "pressed" state.
   */
  pressedTextStyle?: StyleProp<TextStyle>
  /**
   * One of the different types of button presets.
   */
  preset?: Presets
  /**
   * One of the different types of button presets.
   */
  size?: "small" | "medium"
  /**
   * An optional component to render on the right side of the text.
   * Example: `RightAccessory={(props) => <View {...props} />}`
   */
  RightAccessory?: ComponentType<ButtonAccessoryProps>
  /**
   * Icon to display for the `Button`.
   */
  icon?: IconTypes
  /**
   * Size of Icon to display for the `Button`.
   */
  iconSize?: number

  /**
   * Size of Icon to display for the `Button`.
   */
  iconWeight?: IconWeight
  /**
   * An optional style override for the button text.
   */
  iconStyle?: StyleProp<TextStyle>
  /**
   * Whether to show a loading indicator.
   */
  loading?: boolean

  /**
   * Whether to show a full width.
   */
  fullWidth?: boolean
  /**
   * Children components.
   */
  children?: React.ReactNode
}

/**
 * A component that allows users to take actions and make choices.
 * Wraps the Text component with a Pressable component.
 */
export function Pressable(props: ButtonProps): JSX.Element {
  const {
    tx,
    text,
    style: $pressableStyleOverride,
    pressedStyle: $pressedViewStyleOverride,
    viewStyle: $viewStyleOverride,
    pressedTextStyle: $pressedTextStyleOverride,
    children,
    disabled,
    size = "medium",
    ...rest
  } = props

  f()
  function f() {
    return null
  }

  const preset: Presets = $viewPresets[props.preset] ? props.preset : "primary"

  function pressableStyle({ pressed }): StyleProp<ViewStyle> {
    return [
      !!tx || !!text ? $viewPresets[preset][size] : $viewOnlyIconPresets[preset][size],
      $pressableStyleOverride,
      !!pressed && [$pressedViewPresets[preset], $pressedViewStyleOverride],
      !!disabled && [$pressedViewDisabledPresets[preset], $pressedViewStyleOverride],
    ]
  }

  function viewStyle({ pressed }): StyleProp<ViewStyle> {
    return [
      $textPresets[preset],
      $viewStyleOverride,
      !!pressed && [$pressedTextPresets[preset], $pressedTextStyleOverride],
      !!disabled && [$textDisabledPresets[preset], $pressedTextStyleOverride],
    ]
  }

  return (
    <RNPressable style={pressableStyle} {...rest}>
      {(state): ReactNode => (
        <>
          <View style={viewStyle(state)}>{children}</View>
        </>
      )}
    </RNPressable>
  )
}
