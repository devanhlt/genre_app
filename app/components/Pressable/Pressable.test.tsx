/* eslint-disable react-native/no-inline-styles */
/* eslint-disable react-native/no-color-literals */
import React from "react"

import renderer from "react-test-renderer"
import { Pressable } from "./Pressable"

describe("<Pressable />", () => {
  it("renders correctly with isRTL is true", () => {
    const tree = renderer.create(
      <Pressable preset="primary" text="body02 - Cillum eu laboris in labore" />,
    )
    expect(tree).toMatchSnapshot()
  })

  it("renders correctly with text which is looked up via i18n", () => {
    const Component = (
      <>
        <Pressable preset="primary" tx="common.cancel" />

        <Pressable
          pressedStyle={{ backgroundColor: "red" }}
          preset="secondary"
          text="body03 - Cillum eu laboris in labore"
          size="medium"
        />
        <Pressable
          pressedStyle={{ backgroundColor: "red" }}
          preset="secondary"
          text="body03 - Cillum eu laboris in labore"
          size="small"
        />
        <Pressable
          pressedStyle={{ backgroundColor: "red" }}
          preset="primary"
          text="body03 - Cillum eu laboris in labore"
          size="small"
        />

        <Pressable pressedStyle={{ backgroundColor: "red" }} preset="secondary" size="small">
          <></>
        </Pressable>
        <Pressable pressedStyle={{ backgroundColor: "red" }} preset="tertiary" size="small">
          <></>
        </Pressable>

        <Pressable preset="tertiary" text="body04 - Cillum eu laboris in labore" />
      </>
    )
    const tree = renderer.create(Component)
    expect(tree).toMatchSnapshot()
  })
})
