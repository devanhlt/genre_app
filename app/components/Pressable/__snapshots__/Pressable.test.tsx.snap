// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pressable /> renders correctly with isRTL is true 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  preset="primary"
  style={
    [
      [
        {
          "paddingHorizontal": 16,
          "paddingVertical": 12,
        },
        {
          "backgroundColor": "#C5281C",
        },
      ],
      undefined,
      false,
      false,
    ]
  }
>
  <View
    style={
      [
        [
          {
            "flexGrow": 0,
            "flexShrink": 1,
            "marginStart": 4,
            "textAlign": "center",
            "zIndex": 2,
          },
          {
            "fontFamily": "System",
            "fontSize": 14,
            "fontWeight": "600",
            "letterSpacing": 0.15,
            "lineHeight": 20,
          },
          {
            "color": "#FFFFFF",
          },
        ],
        undefined,
        false,
        false,
      ]
    }
  />
</View>
`;

exports[`<Pressable /> renders correctly with text which is looked up via i18n 1`] = `
[
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    preset="primary"
    style={
      [
        [
          {
            "paddingHorizontal": 16,
            "paddingVertical": 12,
          },
          {
            "backgroundColor": "#C5281C",
          },
        ],
        undefined,
        false,
        false,
      ]
    }
  >
    <View
      style={
        [
          [
            {
              "flexGrow": 0,
              "flexShrink": 1,
              "marginStart": 4,
              "textAlign": "center",
              "zIndex": 2,
            },
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "600",
              "letterSpacing": 0.15,
              "lineHeight": 20,
            },
            {
              "color": "#FFFFFF",
            },
          ],
          undefined,
          false,
          false,
        ]
      }
    />
  </View>,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    preset="secondary"
    style={
      [
        [
          {
            "paddingHorizontal": 16,
            "paddingVertical": 12,
          },
          {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#A1ABBB",
            "borderWidth": 1,
          },
        ],
        undefined,
        false,
        false,
      ]
    }
  >
    <View
      style={
        [
          [
            {
              "flexGrow": 0,
              "flexShrink": 1,
              "marginStart": 4,
              "textAlign": "center",
              "zIndex": 2,
            },
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "600",
              "letterSpacing": 0.15,
              "lineHeight": 20,
            },
            {
              "color": "#B80000",
            },
          ],
          undefined,
          false,
          false,
        ]
      }
    />
  </View>,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    preset="secondary"
    style={
      [
        [
          {
            "paddingHorizontal": 16,
            "paddingVertical": 12,
          },
          {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#A1ABBB",
            "borderWidth": 1,
            "paddingHorizontal": 12,
            "paddingVertical": 8,
          },
        ],
        undefined,
        false,
        false,
      ]
    }
  >
    <View
      style={
        [
          [
            {
              "flexGrow": 0,
              "flexShrink": 1,
              "marginStart": 4,
              "textAlign": "center",
              "zIndex": 2,
            },
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "600",
              "letterSpacing": 0.15,
              "lineHeight": 20,
            },
            {
              "color": "#B80000",
            },
          ],
          undefined,
          false,
          false,
        ]
      }
    />
  </View>,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    preset="primary"
    style={
      [
        [
          {
            "paddingHorizontal": 16,
            "paddingVertical": 12,
          },
          {
            "backgroundColor": "#C5281C",
            "paddingHorizontal": 16,
            "paddingVertical": 8,
          },
        ],
        undefined,
        false,
        false,
      ]
    }
  >
    <View
      style={
        [
          [
            {
              "flexGrow": 0,
              "flexShrink": 1,
              "marginStart": 4,
              "textAlign": "center",
              "zIndex": 2,
            },
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "600",
              "letterSpacing": 0.15,
              "lineHeight": 20,
            },
            {
              "color": "#FFFFFF",
            },
          ],
          undefined,
          false,
          false,
        ]
      }
    />
  </View>,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    preset="secondary"
    style={
      [
        [
          {
            "paddingHorizontal": 16,
            "paddingVertical": 12,
          },
          {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#A1ABBB",
            "borderWidth": 1,
            "paddingHorizontal": 8,
            "paddingVertical": 8,
          },
        ],
        undefined,
        false,
        false,
      ]
    }
  >
    <View
      style={
        [
          [
            {
              "flexGrow": 0,
              "flexShrink": 1,
              "marginStart": 4,
              "textAlign": "center",
              "zIndex": 2,
            },
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "600",
              "letterSpacing": 0.15,
              "lineHeight": 20,
            },
            {
              "color": "#B80000",
            },
          ],
          undefined,
          false,
          false,
        ]
      }
    />
  </View>,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    preset="tertiary"
    style={
      [
        [
          {
            "paddingHorizontal": 16,
            "paddingVertical": 12,
          },
          {
            "backgroundColor": "rgba(0, 0, 0, 0)",
            "paddingHorizontal": 8,
            "paddingVertical": 8,
          },
        ],
        undefined,
        false,
        false,
      ]
    }
  >
    <View
      style={
        [
          [
            {
              "flexGrow": 0,
              "flexShrink": 1,
              "marginStart": 4,
              "textAlign": "center",
              "zIndex": 2,
            },
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "600",
              "letterSpacing": 0.15,
              "lineHeight": 20,
            },
            {
              "color": "#B80000",
            },
          ],
          undefined,
          false,
          false,
        ]
      }
    />
  </View>,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    preset="tertiary"
    style={
      [
        [
          {
            "paddingHorizontal": 16,
            "paddingVertical": 12,
          },
          {
            "backgroundColor": "rgba(0, 0, 0, 0)",
          },
        ],
        undefined,
        false,
        false,
      ]
    }
  >
    <View
      style={
        [
          [
            {
              "flexGrow": 0,
              "flexShrink": 1,
              "marginStart": 4,
              "textAlign": "center",
              "zIndex": 2,
            },
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "600",
              "letterSpacing": 0.15,
              "lineHeight": 20,
            },
            {
              "color": "#B80000",
            },
          ],
          undefined,
          false,
          false,
        ]
      }
    />
  </View>,
]
`;
